<img style="float: right; width: 150px;" src="https://raw.githubusercontent.com/firrm/DAI/main/assets/firrm.jpg">

## <span style="color:#4375c7">DAI</span>
***
*Course materials are for educational purposes only. Nothing contained herein should be considered investment advice or an opinion regarding the suitability of any security. For more information about this course, please contact us.*
***

## Introduction to the Scientific Toolbox of Python: Pandas

Pandas (Python Data Analysis Library) is an open source library for Python that provides powerful, easy-to-use data structures and data analysis tools. It is one of the most popular and widely used tools for data scientists to manipulate, analyze, clean, and prepare data. It is often thought of as the Python version of Microsoft Excel. It should already be installed in your Python environment. If not, install it by typing `pip install pandas` at the command prompt. Test that pandas is installed correctly by running the following line of code:

In [1]:
!pip install -r https://raw.githubusercontent.com/firrm/DAI/main/requirements.txt #ensure that the required packages are installed
# Importing the pandas module as pd (shortened)
import pandas as pd

If no error occurs, pandas is successfully installed in your environment. Now let us analyze data!

### Session contents:
1. **[Pandas Series](#series)**
2. **[Pandas DataFrame](#dataframe)**
3. **[Filtering Data](#filtering)**
4. **[Reading and Analyzing .csv Data](#csv)**
5. **[Merging DataFrames](#merge)**
6. **[Missing Data](#missing)**
7. **[Hands-On](#hands-on)**
8. **[Session Takeaways](#takeaways)**
***

## 1. Pandas Series <a name="series"></a> 

What is a Pandas series? A series is a one-dimensional labeled array that stores data. Because it has only one dimension, it stores only one column of information. Think of it as a basic data type: it can store any type of data (integers, booleans, etc.), but ideally it stores only *one* type of data. Therefore, be consistent when creating series. To construct a series, call the `series' function, which creates a new object of type series.

Recall the list of BMW's cash flows that we analyzed in the previous lecture.

In [2]:
# Fictive cash flow of BMW
BMW = [300, 150, 125, 270, 300, 350, 325, 340, 360, 350, 355]

# Let us create a series of the above list and print it out
s1 = pd.Series(BMW)
print(s1)

0     300
1     150
2     125
3     270
4     300
5     350
6     325
7     340
8     360
9     350
10    355
dtype: int64


Note that `dtype` indicates the data type of the series. If the series contains strings, Pandas will flag that the data type is `object`. 

The first column shows the indices of the values, and like Python lists, starts with index 0. An advantage over Python lists is that the indices do not have to be numeric values, but can be any data type. To illustrate this, consider the following list of quarterly periods that refer to the notional cash flows:

In [3]:
quarters = ['q1-2017', 'q2-2017', 'q3-2017', 'q4-2017', 
           'q1-2018', 'q2-2018', 'q3-2018', 'q4-2018', 
           'q1-2019', 'q2-2019', 'q3-2019']

We now want to add the quarters as an index column to the notional cash flows.

In [4]:
# Create a new series
s2 = pd.Series(data=BMW, index=quarters)
# alternatively type: s2=pd.Series(BMW,quaters)

s2

q1-2017    300
q2-2017    150
q3-2017    125
q4-2017    270
q1-2018    300
q2-2018    350
q3-2018    325
q4-2018    340
q1-2019    360
q2-2019    350
q3-2019    355
dtype: int64

We now have a labeled index series! Instead of numeric values starting with 0, the series is indexed by strings. 

In general, we can build a Pandas-based series using a tuple, an array, or a dictionary, but we cannot use a set, because a set is unordered. Keep this in mind!

To specifically retrieve information from the Pandas series, we use the index or label to get the corresponding value, e.g:

In [5]:
s2['q1-2019']

360

We can also perform arithmetric operations with Pandas series. Pandas will match the indexes or labels in the series and perform the operation on the respective values. For labels that do not match, Pandas will put a null value or a NaN, which means "missing value" (`Null = NaN = Missing Value`): 

In [6]:
# Example for adding s1 and s2
s1 + s2

0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
5         NaN
6         NaN
7         NaN
8         NaN
9         NaN
10        NaN
q1-2017   NaN
q1-2018   NaN
q1-2019   NaN
q2-2017   NaN
q2-2018   NaN
q2-2019   NaN
q3-2017   NaN
q3-2018   NaN
q3-2019   NaN
q4-2017   NaN
q4-2018   NaN
dtype: float64

As we can see above, the labels do not match, so pandas returns NaN values. If we have series with matching indices, the calculation will work.

## 2. Pandas DataFame <a name="dataframe"></a> 

According to the Pandas documentation, a data frame is a two-dimensional, mutable, and potentially heterogeneous tabular data structure with labeled axes (rows and columns). Arithmetic operations are applied to both row and column labels. A data frame is the primary data structure of Pandas.

In fact, a data frame is a list of lists, and the first element is a list. For example, consider stock prices. If we want to display a company and its opening and closing stock prices, we can create the following data frame:

In [7]:
# Goal: presenting this list of stock prices in a data frame

stock_prices = [['Adidas', 30, 31], ['BMW', 40, 36],
               ['Dt Bank', 12, 10], ['SAP', 70, 76]]

# Creating a pandas data frame object
prices_df = pd.DataFrame(data=stock_prices, columns=['Company', 'Open', 
                                                     'Close'])

So, what is the type of `prices_df`? And what is the type of a single column within the data frame?

In [8]:
type(prices_df), type(prices_df['Company'])

(pandas.core.frame.DataFrame, pandas.core.series.Series)

Using the shape attribute, we can analyze the number of rows and columns of the above data frame:

In [9]:
prices_df.shape

(4, 3)

Now, print out the Pandas data frame that we just created.

In [10]:
prices_df

Company  Open  Close
0   Adidas    30     31
1      BMW    40     36
2  Dt Bank    12     10
3      SAP    70     76

Note that both rows and columns are Pandas series. Using the `info` command, we can gain more useful insights on the data frame:

In [11]:
prices_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Company  4 non-null      object
 1   Open     4 non-null      int64 
 2   Close    4 non-null      int64 
dtypes: int64(2), object(1)
memory usage: 224.0+ bytes


Since most of the time we are interested in the closing prices, we want to modify our data frame and select only the `Company` and `Close`. Interestingly, most of the trading happens towards the end of the trading day, so the closing prices reflect the last set of information and the realized trades of the day.

In [12]:
# Modifying the prices data frame
closing_prices = prices_df[['Company','Close']]
closing_prices

Company  Close
0   Adidas     31
1      BMW     36
2  Dt Bank     10
3      SAP     76

We can accomplish the same modification by using the `.drop(x,y)` function for Pandas data frames, where `x` stands for the index or column name and `y` refers to the axis (row=0 and column = 1): 

In [13]:
prices_df.drop('Open', axis=1)

Company  Close
0   Adidas     31
1      BMW     36
2  Dt Bank     10
3      SAP     76

So far, we have focused on creating and manipulating columns in data frames. Now let's look at how to handle rows. Basically, there are two methods for selecting rows: `loc[]` and `iloc[]`. When using these methods, we pass the label and and it will return a row that is actually a series. For example, return the row entries for index 2:

In [14]:
# Return the row entries of prices_df for index 2
prices_df.loc[2]

Company    Dt Bank
Open            12
Close           10
Name: 2, dtype: object

Intersections of rows and columns are found by using `.loc[x,y]`, where `x` is the index and `y` denotes the column name. Print out the closing price of SAP:

In [15]:
prices_df.loc[3,'Close']

76

To select multiple rows at once, we can use slicing, e.g., to select rows 0 to 2 from the price data frame, we can type the following:

In [16]:
prices_df.loc[0:2, ['Company', 'Close']]

Company  Close
0   Adidas     31
1      BMW     36
2  Dt Bank     10

Note that the second argument of the `.loc` function is a list. 

Above we also mentioned a second method called `.iloc[]`. This method uses index-based slicing, which does not include the stop index. The main difference between `.loc` and `.iloc` is that `.loc` uses labels and `.iloc` always uses indices starting from zero ("Python internal counting"). For example, if we change the index that pandas automatically used and use strings as indices, we will still give the `.iloc` method 0 as an argument to target the first row. To illustrate, let us make an example:

In [17]:
# Printing only the closing price of BMW
prices_df.iloc[0,2]

31

For statistical methods, it is often relevant to simulate a random draw from a sample. Given a data frame, the function `.sample(n)` with `n` as the number of rows does the job.

In [18]:
prices_df.sample(n=2)

Company  Open  Close
3     SAP    70     76
1     BMW    40     36

## 3. Filtering Data <a name="filtering"></a> 

Most of the time, data is messy. To prepare the data for statistical and econometric analysis, we need to clean up the data sets, and filtering is an essential method for doing so. Filtering means selecting a subset in our data frame based on a certain condition that should be met. For now, we will work with the above data frame:

In [19]:
prices_df

Company  Open  Close
0   Adidas    30     31
1      BMW    40     36
2  Dt Bank    12     10
3      SAP    70     76

Let us filter this data set based on the condition that the closing price should be above 35:

In [20]:
prices_df['Close']>35

0    False
1     True
2    False
3     True
Name: Close, dtype: bool

This is in fact a series of booleans. Now let us select and return all rows where the closing price is above 35. This is a condition based on a column. So we write:

In [21]:
prices_df[prices_df['Close'] > 35]

Company  Open  Close
1     BMW    40     36
3     SAP    70     76

It returns all rows where the closing price is greater than 35. This also works with the `.loc` method we learned earlier, e.g:

In [22]:
prices_df[prices_df.loc[:, 'Close'] > 35]

Company  Open  Close
1     BMW    40     36
3     SAP    70     76

We can further filter the output by specifying the column that shall be returned. Note that for multiple columns, you must create a list of lists.

In [23]:
# Only print out the companies that fulfill the condition
prices_df[prices_df.loc[:, 'Close'] > 35]['Company']

1    BMW
3    SAP
Name: Company, dtype: object

If you want to filter based on multiple conditions, put each condition in parentheses and use the `&` symbol (i.e., logical `and`) or a vertical bar `|` (i.e., for logical `or`) to specify your condition more precisely. 

Now we want to calculate the maximum and minimum of all values in a column. What is the maximum/minimum closing price?

In [24]:
prices_df['Close'].max()

76

In [25]:
prices_df['Close'].min()

10

In [26]:
# Reloading the stock price data frame
stock_prices = [['Adidas', 30, 31], ['BMW', 40, 36],
               ['Dt Bank', 12, 10], ['SAP', 70, 76]]

# Creating a pandas data frame object
prices_df = pd.DataFrame(data=stock_prices, columns=['Company', 'Open', 
                                                     'Close'])
prices_df

Company  Open  Close
0   Adidas    30     31
1      BMW    40     36
2  Dt Bank    12     10
3      SAP    70     76

Another important feature is to change the index. In a data frame setting, this can be done as follows:

In [27]:
prices_df.set_index('Company', inplace = True)

In [28]:
prices_df

Open  Close
Company             
Adidas     30     31
BMW        40     36
Dt Bank    12     10
SAP        70     76

Finally, the `between` method is useful when you want to find values that fall within a certain range. For example, filter for all companies that have a closing price between 5 and 35.

In [29]:
prices_df[prices_df['Close'].between(5,35)]

Open  Close
Company             
Adidas     30     31
Dt Bank    12     10

Note that the arguments of the `between` method are inclusive!

## 4. Reading and Analyzing .csv Data <a name="csv"></a> 

In this section, we will learn how to import data into our project. In fact, this is the first step of any data science / statistical / econometric application. Pandas as a library can import data from a variety of sources, such as .csv, .json, Excel files, .html pages, or SQL databases. Here we will focus on working with data from .csv files. Since the Pandas library is already imported, we can directly read in our CSV-file of s&p500 stock prices from 2000 to 2019.

In [30]:
# Reading in the csv file into a pandas DataFrame
sp500 = pd.read_csv('https://raw.githubusercontent.com/firrm/DAI/main/data/sp500.csv', decimal = '.', sep = ',')
sp500 = pd.DataFrame(sp500)

# Check the properties of the newly created data frame
sp500.info()
sp500.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4945 entries, 0 to 4944
Columns: 506 entries, date to NKTR
dtypes: float64(505), object(1)
memory usage: 19.1+ MB


(4945, 506)

Let us display the data frame. 

Note that not all rows are displayed, as the browser may freeze if the data frame is too large. Using `pd.set_option()` we can specify how many rows we want to display.

In [31]:
sp500

date       MSFT        AAPL       AMZN     GOOG    GOOGL      FB  \
0     2000-01-03   58.28125    3.997764    89.3750      NaN      NaN     NaN   
1     2000-01-04   56.31250    3.660711    81.9375      NaN      NaN     NaN   
2     2000-01-05   56.90625    3.714282    69.7500      NaN      NaN     NaN   
3     2000-01-06   55.00000    3.392854    65.5625      NaN      NaN     NaN   
4     2000-01-07   55.71875    3.553568    69.5625      NaN      NaN     NaN   
...          ...        ...         ...        ...      ...      ...     ...   
4940  2019-08-22  137.78000  212.460000  1805.6000  1189.53  1191.52  182.04   
4941  2019-08-23  133.39000  202.640000  1749.6200  1151.29  1153.58  177.75   
4942  2019-08-26  135.45000  206.490000  1768.8700  1168.89  1171.18  180.36   
4943  2019-08-27  135.74000  204.160000  1761.8300  1167.84  1170.82  181.30   
4944  2019-08-28  135.56000  205.530000  1764.2500  1171.02  1173.75  181.76   

       BRK.B       V        JNJ  ...        PWR         LB         M  \
0      35.30     NaN   46.09375  ...  18.791667  19.018868  25.18750   
1      34.08     NaN   44.40625  ...  18.750000  18.637904  24.46875   
2      34.64     NaN   45.03125  ...  19.041667  18.374160  24.93750   
3      36.08     NaN   46.40625  ...  19.083333  18.081111  24.71875   
4      36.40     NaN   48.25000  ...  18.583333  18.198331  25.40625   
...      ...     ...        ...  ...        ...        ...       ...   
4940  201.01  180.09  131.27000  ...  33.480000  19.330000  15.56000   
4941  197.16  175.23  127.73000  ...  33.410000  17.530000  14.94000   
4942  200.20  177.59  127.80000  ...  33.210000  16.920000  14.71000   
4943  198.35  178.38  129.64000  ...  32.660000  16.150000  14.30000   
4944  200.45  178.67  128.68000  ...  32.970000  16.670000  14.62000   

            JWN         HP   CPRI        MAC    XEC        AMG       NKTR  
0     12.022752   7.550513    NaN  17.911277    NaN  24.499755  20.843750  
1     11.555849   7.321710    NaN  17.580607    NaN  22.541441  22.000000  
2     11.351579   7.321710    NaN  17.856165    NaN  21.999780  22.015625  
3     11.906026   7.756437    NaN  18.186835    NaN  22.666440  22.843750  
4     11.906026   7.939479    NaN  18.903286    NaN  23.208101  23.000000  
...         ...        ...    ...        ...    ...        ...        ...  
4940  30.750000  39.040000  28.01  28.860000  40.21  76.510000  16.980000  
4941  28.570000  37.060000  26.75  27.880000  38.11  74.640000  16.990000  
4942  28.850000  37.080000  26.74  28.040000  38.12  75.080000  17.290000  
4943  27.890000  36.650000  25.89  27.760000  37.63  74.840000  16.910000  
4944  28.700000  36.820000  26.20  28.280000  41.60  75.030000  17.460000  

[4945 rows x 506 columns]

Now, we set the timestamp as the index. 

In [32]:
sp500.set_index('date', inplace = True)


Now we want to drop companies for which we do not have stock quotes. We use `dropna()` to drop these columns from the s&p500 data frame. We then check the dimension of the data frame to verify the filtering operation. Finally, we print the data frame.

In [33]:
sp500 = sp500.dropna()
sp500

MSFT    AAPL     AMZN     GOOG     GOOGL      FB   BRK.B  \
date                                                                     
2019-06-11  132.10  194.81  1863.70  1078.72  1081.040  178.10  206.63   
2019-06-12  131.49  194.19  1855.32  1077.03  1079.100  175.04  204.57   
2019-06-13  132.32  194.15  1870.30  1088.77  1091.010  177.47  204.99   
2019-06-14  132.45  192.74  1869.67  1085.35  1086.300  181.33  205.16   
2019-06-17  132.85  193.89  1886.03  1092.50  1093.890  189.01  203.48   
2019-06-18  135.16  198.45  1901.37  1103.60  1105.240  188.47  206.01   
2019-06-19  135.69  197.87  1908.79  1102.33  1104.510  187.48  206.75   
2019-06-20  136.95  199.46  1918.19  1111.42  1113.200  189.53  208.04   
2019-06-21  136.97  198.78  1911.30  1121.88  1125.370  191.14  206.21   
2019-06-24  137.78  198.58  1913.90  1115.52  1116.700  192.60  206.85   
2019-06-25  133.43  195.57  1878.27  1086.35  1087.580  188.84  208.07   
2019-06-26  133.93  199.80  1897.83  1079.80  1080.320  187.66  208.51   
2019-06-27  134.15  199.74  1904.28  1076.01  1076.630  189.50  212.02   
2019-06-28  133.96  197.92  1893.63  1080.91  1082.800  193.00  213.17   
2019-07-01  135.68  201.55  1922.19  1097.95  1100.000  193.00  214.62   
2019-07-02  136.58  202.73  1934.31  1111.25  1112.600  195.00  214.54   
2019-07-03  137.46  204.41  1939.00  1121.58  1122.990  197.20  215.16   
2019-07-05  137.06  204.23  1942.91  1131.59  1132.665  196.40  215.25   
2019-07-08  136.96  200.02  1952.32  1116.35  1116.790  195.76  215.13   
2019-07-09  136.46  201.24  1988.30  1124.83  1124.290  199.21  214.75   
2019-07-10  137.85  203.23  2017.41  1140.48  1140.910  202.73  214.03   
2019-07-11  138.40  201.75  2001.07  1144.21  1144.080  201.23  213.17   
2019-07-12  138.90  203.30  2011.00  1144.90  1145.340  204.87  214.10   
2019-07-15  138.90  205.21  2020.99  1150.34  1150.510  203.91  213.54   
2019-07-16  137.08  204.50  2009.90  1153.58  1153.460  203.84  212.88   
2019-07-17  136.27  203.35  1992.03  1146.35  1146.740  201.80  207.45   
2019-07-18  136.42  205.66  1977.90  1146.33  1147.240  200.78  207.98   
2019-07-19  136.62  202.59  1964.52  1130.10  1131.550  198.36  205.97   
2019-07-22  138.43  207.22  1985.63  1138.07  1139.210  202.32  204.84   
2019-07-23  139.29  208.84  1994.49  1146.21  1148.050  202.36  206.46   
2019-07-24  140.72  208.67  2000.81  1137.81  1139.730  204.66  207.97   
2019-07-25  140.19  207.02  1973.82  1132.12  1135.940  200.71  207.90   
2019-07-26  141.34  207.74  1943.05  1250.41  1245.220  199.75  209.90   
2019-07-29  141.03  209.68  1912.45  1239.41  1241.840  195.94  208.61   
2019-07-30  140.35  208.78  1898.53  1225.14  1228.000  197.04  207.81   
2019-07-31  136.27  213.04  1866.78  1216.68  1218.200  194.23  205.43   
2019-08-01  138.06  208.43  1855.32  1209.01  1211.780  192.73  201.71   
2019-08-02  136.90  204.02  1823.24  1193.99  1196.320  189.02  202.67   
2019-08-05  132.21  193.34  1765.13  1152.32  1154.750  181.73  197.73   
2019-08-06  134.69  197.00  1787.83  1169.95  1171.080  184.51  200.55   
2019-08-07  135.28  199.04  1793.40  1173.99  1175.910  185.15  198.21   
2019-08-08  138.89  203.43  1832.89  1204.80  1206.190  190.16  200.52   
2019-08-09  137.71  200.99  1807.58  1188.01  1188.900  187.85  199.05   
2019-08-12  135.79  200.48  1784.92  1174.71  1174.500  185.37  197.20   
2019-08-13  138.60  208.97  1824.34  1197.27  1196.730  188.45  198.78   
2019-08-14  133.98  202.75  1762.96  1164.29  1164.250  179.71  195.81   
2019-08-15  133.68  201.74  1776.12  1167.26  1169.320  182.59  198.31   
2019-08-16  136.13  206.50  1792.57  1177.60  1179.210  183.70  199.72   
2019-08-19  138.41  210.35  1816.12  1198.45  1200.440  186.17  200.94   
2019-08-20  137.26  210.36  1801.38  1182.69  1183.530  183.81  198.15   
2019-08-21  138.79  212.64  1823.54  1191.25  1191.580  183.55  199.24   
2019-08-22  137.78  212.46  1805.60  1189.53  1191.520  182.04  201.01   
2019-08-23  133.39

## 5. Merging DataFrames <a name="merge"></a> 

**Short homework**: Familiarize yourself with merging different data frames. This can be very useful when dealing with many different data sources that need to be merged into one large data set. This [link](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html) gives you access to an extensive tutorial, but feel free to explore on your own and/or use DataCamp to review the basics of merging.

## 6. Missing Data <a name="missing"></a> 

Since data comes in different forms, we will now see how to work with missing data. Pandas is generally flexible in its handling of missing values. In Pandas, missing values are referred to as NaN = zero = missing data. Imagine there is a column that only exists in one data set. If we join this data set with another data set, NaNs will be introduced.

As we have seen in the `sp500` dataframe, some companies had NaN values in their price history (so we dropped them). In general, a useful method to drop NaNs is `df.dropna()`. If there are only a few missing values, we simply drop them (assuming our sample is large enough). If we call this method without any arguments, it will drop all rows that have at least one missing value. Note that this command does not modify the existing data frame, but simply returns a newly modified data frame. By default, the method removes missing values across rows, so it removes rows with missing values. If we decide to remove columns with one or more missing values, we use `df.dropna(axis=1)`, which is the default for columns.

Another useful argument of `df.dropna()` is the threshold option `thresh`. It will drop rows or columns that have less than the threshold number of non-NaN values. For example, if the threshold is 5, it will keep all rows/columns that have at least 5 known values.

`fillna()` is a method used to fill missing values in our data frame. It replaces the NaN values with something else that we specify.

## 7. Hands-On <a name="hands-on"></a> 

In Section 4, we successfully created a data frame with the stock prices of the 506 constituents of the US S&P500 index. We now want to analyze the return distributions of the different stocks.

**a)** Calculate the returns for each stock as:

\begin{align}
r_t=\frac{P_t}{P_{t-1}}-1
\end{align}

and save the returns in a dataframe `sp500_returns`.

**b)** The skewness - or third central moment - is defined as

\begin{align}
S=E \left[ \left( \frac{X-\mu}{\sigma} \right) ^3 \right]
\end{align}

and the kurtosis - or fourth central moment - is defined as

\begin{align}
K=E \left[ \left( \frac{X-\mu}{\sigma} \right) ^4 \right]
\end{align}

Write functions in python for both statistics and calculate the skewness and kurtosis of the return distributions of the 506 constituents of the S&P500. Save your results in a data frame called `sp500_statistics`.

**c)** Regarding distributional properties, return series are often assumed to be normally distributed. To verify this, write a function in python that calculates the Jarque-Bera Statistic for each return distribution:

\begin{align}
JB=\frac{n}{6} \left( S^2 + \frac{(K-3)^2}{4} \right)
\end{align}

where $n$ is the number of observations, $S$ is the skewness and $K$ is the kurtosis. Save your results in a data frame. How can you interpret your results?

## 8. Session Takeaways <a name="takeaways"></a> 

*What have we learned in this session?*

- Introduction to Pandas library
- How to create and manipulate Pandas series and dataframes
- How to import data from csv files
- How to merge data frames (as homework)
- How to deal with missing data in our data set
- Import stock data, conduct data frame filtering manipulations, calculate returns and analyze statistical properties of distributional features

*What's next?*

We are now familiar with the first scientific library of the Python 3 programming language. In the following, we will look at the Matplotlib and Scipy libraries, which will enhance our statistical and econometric analysis capabilities.

***

To practice the pandas functions, please take the datacamp courses "Manipulating DataFrames with Pandas","Merging DataFrames with Pandas" and "Cleaning Data in Python".
For a quick overview of the pandas package, check: http://datacamp-community-prod.s3.amazonaws.com/dbed353d-2757-4617-8206-8767ab379ab3